validator/src/proposal_buffer.rs

```diff
@@ -0,0 +1,340 @@
+        if shared.buffer.is_empty() {
+            store_waker!(shared, waker, cx);
+            return Poll::Pending;
+        }
+
+        let proposal_and_pubsubid = shared.buffer.entries().find_map(|entry| {
+            let (signed_proposal, _pubsub_id) = entry.get();
+
+            // Get a read lock of the blockchain.
+            let blockchain = self.blockchain.read();
+
+            // Try and retrieve the predecessor of the proposal.
+            match blockchain.get_block(&signed_proposal.proposal.parent_hash, false, None) {
+                Ok(Block::Micro(block)) => {
+                    // Proposal predecessor is known.
+                    // Proceed to make sure the signer, whose signature was already verified,
+                    // is in fact also the one who was supposed to produce this proposal.
+
+                    // Get the active validators.
+                    let validators = self.blockchain.read().current_validators().unwrap();
+
+                    // Get the validator who signed the proposal and whose signature was already verified.
+                    let assumed_validator =
+                        validators.get_validator_by_slot_band(signed_proposal.signer);
+
+                    // Calculate the validator who was supposed to produce the block.
+                    let actual_validator = blockchain
+                        .get_proposer_at(
+                            signed_proposal.proposal.block_number,
+                            signed_proposal.round,
+                            block.header.seed.entropy(),
+                            None,
+                        )
+                        .expect("Must be able to calculate block producer.")
+                        .validator;
+
+                    // Compare the expected and the actual validator
+                    if *assumed_validator == actual_validator {
+                        // Validators match. The signature was already verified. Proposal is good to go.
+                        // DO NOT validate message the proposal, as there is much more to the validity than the signature.
+                        Some(entry.remove())
+                    } else {
+                        // Assumed validator is not the proposer. Punish the validator (todo).
+                        // Proposal is removed as it is not a valid proposal.
+                        log::debug!("Found MacroHeader Proposal where validators do not match.");
+                        let (_proposal, id) = entry.remove();
+
+                        // The peer relaying this proposal relayed a faulty proposal, disconnect him.
+                        let source = id.propagation_source();
+                        let nw = Arc::clone(&self.network);
+                        tokio::task::spawn(async move {
+                            nw.disconnect_peer(source, CloseReason::MaliciousPeer).await;
+                        });
```

Comment:
Theoretically, we're already in an `async` function, so no need to `tokio::spawn` another task for this. A list of disconnect-peer-futures could be stored in the `ProposalReceiver` and polled every time `poll_next` is called.

Might be too annoying to implement though. 🤷‍♀️